README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/villano-lab/galactic-spin-W1/HEAD?labpath=binder)
[![TravisCI Build Status](https://app.travis-ci.com/villano-lab/galactic-spin-W1.svg?branch=master)](https://app.travis-ci.com/github/villano-lab/galactic-spin-W1)
[![CircleCI Build Status](https://dl.circleci.com/status-badge/img/gh/villano-lab/galactic-spin-W1/tree/master.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/villano-lab/galactic-spin-W1/tree/master)
[![Documentation Status](https://readthedocs.org/projects/galactic-spin-w1/badge/?version=stable)](https://galactic-spin-w1.readthedocs.io/en/latest/?badge=stable)
[![Codecov](https://codecov.io/gh/villano-lab/galactic-spin-W1/branch/master/graph/badge.svg?token=U7YINMN0V4)](https://codecov.io/gh/villano-lab/galactic-spin-W1)

# Galactic Spin W1: The Basics of Rotation Curves

[![JOSE status](https://jose.theoj.org/papers/9fc3443a7c138ea1a3e1a15f4a3f14df/status.svg)](https://jose.theoj.org/papers/9fc3443a7c138ea1a3e1a15f4a3f14df)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.6588350.svg)](https://doi.org/10.5281/zenodo.6588350)<br/>

The first workshop of our Rotation Curves series, which covers the basics of what rotation curves are,
how they are measured, and how we model them by breaking them into components.

You can find more detailed documentation about the code involved in this workshop [here](https://galactic-spin-W1.readthedocs.io/en/stable), 
including [a page guiding you through the workshop itself](https://galactic-spin-w1.readthedocs.io/en/stable/01_Getting_Started.html).

## ABSTRACT

Dark matter is estimated to make up ~84% of all normal/baryonic matter, but cannot be directly imaged. 
Despite the fact that dark matter cannot be directly observed yet, its influence on the motion of stars and gas in spiral galaxies have been detected. 
One way to show motion in galaxies are rotation curves that are plots of velocity measurements of how fast stars and gas move in a galaxy around the center of mass. 
According to Newton's Law of Gravitation, the rotational velocity is an indication of the amount of visible and non-visible mass in the galaxy. 
Given that the visible matter is measurable using photometry, dark matter mass can therefore be estimated, offering an insight into the size distribution in galaxies. 
In order to gain a greater appreciation of the research scientists' findings about dark matter, their method should be easily reproduced by any curious individual. 
Our interactive workshop is an excellent educational tool to investigate how dark matter impacts the rotation of visible matter by providing a guide to produce galactic rotation curves. 
The Python-based notebooks are set up to walk you through the whole process of producing rotation curves using an online database (SPARC) and to allow you to learn about each component of the galaxy. 
The three steps of the rotation curve building process is plotting the measured velocity data, constructing the rotation curves for each component, and fitting the total velocity to the measured values. 

## OUTLINE

| Module Name                                                                | Notebook Name                         | Description |
| -------------------------------------------------------------------------- | ------------------------------------- | ----------- |
| Understanding Rotation Curves                                              | 01_DM_Rotation_Curve_Intro.ipynb       | Rotation curves are a measure of how fast an object is moving at a certain distance from the center. Three kinds of rotation curves are discussed in this notebook: rigid-body rotation, planetary rotation, and galactic flat rotation. With the help of animations and rotation curve plots, the three cases are compared to gain a better understanding of galactic rotation curves. |
| Introduction to dark matter                                                | 02_Widget_NGC5533_DMonly.ipynb         | Mass in a galaxy can be visualized by plotting its rotation curve. A galactic rotation curve is a measure of how fast the stars and gas move in the galaxy at a certain distance from the center. According to Newton's law of gravitation, objects orbiting the center of gravity should depend on the mass enclosed in the system. However, the theoretical rotation curve of the measured visible matter does not agree with the measured velocities of matter in some of the spiral galaxies. Dark matter is introduced to account for the "missing matter". |
| Plotting measured velocities                                               | 03_Measured_Data_Plotting.ipynb        | Plotting measured velocities	3_Measured_Data_Plotting.ipynb	First step in understanding a rotation curve is to plot it. Comparing the rotation curves of multiple galaxies is a good exercise to visualize the motions of stars and gas. The shapes of curves reveal the mass distribution in a galaxy. For example, higher velocity measurements in the central region indicates a supermassive black hole at the center of that galaxy. A flat rotation curve suggests the presence of a dark matter halo. |
| Plotting the components of galactic rotation curves (4 galaxies)           | 04_Plotting_Rotation_Curves.ipynb      | Theoretical rotation curves are computed using the velocities of each component of the galaxy. The velocities of the bulge, disk and gas are calculated from luminosities, surface brightness profiles, surface density profiles or mass models but these calculations are beyond the scope of this workshop. For this reason, the velocities of the three components are imported into the notebook. On the contrary, the rotation curve of the central black hole (point-mass rotation curve - yet another type of rotation curve) and the dark matter halo can be easily produced. Adding all components, the total velocity of only light matter can then be compared to the total velocity that includes the dark matter component. Is it possible to fit both curves to the measured data? |
| Interactive widget to visualize the components of NGC 5533                  | 05_Widget_NGC5533_All_Components.ipynb | The rotation curves of each component in the galaxy NGC 5533 can be scaled up and down using the interactive widget in this notebook. The best combination of the scaling parameters results in a good fit to the measured data points. To characterize the goodness of the fit, the reduced chi-squared value is calculated. A value close to 1 is an indication of a good fit. |
| Plotting the components of galactic rotation curves (SPARC - 175 galaxies)  | 6_Plotting_SPARC_Data.ipynb           | The Spitzer Photometry & Accurate Rotation Curves (SPARC) database provides pre-calculated velocities of the bulge, disk and gas in 175 galaxies, as well as the measured velocity data points. Analogous to the 4_Plotting_Rotation_Curves.ipynb activity, this notebook also gives a guide to calculating the missing dark matter component and compares the rotation curve of luminous matter to the total velocity with the dark matter component. |
| Creating a rotation curve of the bulge component                            | 07_Bonus_Bulge_Rotation_Curve.ipynb    | The bulge is the most luminous, central component of a spiral galaxy that contains densely packed stars and gas. Although the rotation curve of the bulge can be derived from luminosity measurements, more theoretical models only utilize empirically-derived parameters such as the central surface brightness, the total luminosity of the bulge, the concentration parameter, and a characteristic radius. With the use of these parameters and calculus, the theoretical rotation of the bulge can be derived. |
| Interactive fitting of 4 galaxies                                           | 08_Interactive_Fitting.ipynb           | Fitting is a statistical method to scale parameters of a function until it closely resembles the curve of data points. Selecting a galaxy out of four options, these free parameters can be adjusted to vary. The results of the fitting is shown and explained in this notebook. |
| Interactive widget to visualize the components of multiple galaxies (SPARC) | 09_Widget_SPARC_Galaxies.ipynb         | After selecting a galaxy from the SPARC database of 175 galaxies, each component can be scaled up and down using the sliders of the interactive widget. Additionally, the fit parameters of the best fit to the measured velocities is revealed, along with an image of the selected galaxy from the NASA SkyView database. |
| Tiny black holes as dark matter candidates                                  | 10_Bonus_Black_Holes_as_DM.ipynb      | How many tiny black holes can account for the missing mass called dark matter? The interactive widgets and the visual representations of the number and mass of black holes give a good explanation to this question. For comparison, two spiral galaxies are investigated: NGC 5533 and NGC 7814. |

## CITATION

If you decide to use this code, or if you want to add a reference to it, please cite the latest archived version,

> Villano, A.N., Bergfalk, J., Hatami, R., Harris, K., Vititoe, F., and Johnston, J., 2023, The Data Behind Dark Matter: Exploring Galactic Rotation [Code, v3.0.5] [DOI:10.5281/zenodo.6588350].

```
@misc{villano_a_n_2023_6588350,
  author       = {Villano, A. N. and
                  Harris, Kitty C. and
                  Bergfalk, Judit and
                  Hatami, Raphael and
                  Vititoe, Francis and
                  Johnston, Julia},
  title        = {{The Data Behind Dark Matter: Exploring Galactic 
                   Rotation}},
  month        = August,
  year         = 2023,
  publisher    = {Zenodo},
  version      = {v3.0.5},
  doi          = {10.5281/zenodo.6588350},
  url          = {https://doi.org/10.5281/zenodo.6588350}
}

```

## VERSION HISTORY

08.08.2023: Release of [version 3.0.5](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.5)
19.05.2023: Release of [version 3.0.4](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.4)  
18.05.2023: Release of [version 3.0.3](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.3)  
17.05.2023: Release of [version 3.0.2](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.2)  
07.04.2023: Release of [version 3.0.1](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.1)  
28.07.2022: Release of [version 3.0.0](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v3.0.0)  
30.06.2022: Release of [version 2.0.1](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v2.0.1)  
24.06.2022: Release of [version 2.0.0](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v2.0.0)  
17.06.2022: Release of [version 1.0.4](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v1.0.4)  
17.06.2022: Release of [version 1.0.3](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v1.0.3)  
09.06.2022: Release of [version 1.0.2](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v1.0.2)  
27.04.2022: Release of [version 1.0.1](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v1.0.1)  
27.04.2022: Release of [version 1.0.0](https://github.com/villano-lab/galactic-spin-W1/releases/tag/v1.0.0)

## AUTHORS & CONTACT

The authors of *Galactic Spin W1* are A.N. Villano, Kitty C. Harris, Raphael Hatami, Judit Bergfalk, Francis Vititoe, and Julia Johnston.

For questions, support, bug reports, or other suggestions, please open an [issue](https://github.com/villano-lab/galactic-spin-W1/issues).

## LICENSE
This work is licensed under an
[MIT License][mit].

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

[mit]: https://mit-license.org/

Or see the [LICENSE file](https://github.com/villano-lab/galactic-spin-W1/blob/master/LICENSE).


RELEASENOTES.md
## Patch (v3.0.5) Date 08.08.23
Final fixes for JOSE publication
* Grammatical fixes
* Fixes reference formatting
* Add CITATION.cff

## Hotfix (v3.0.4) Date 19.05.23
Fix abstract.
* Fix inaccuracy regarding amount of dark matter.
* Remove the (now erroneous) statement that abstract is from RaCAS 2022.

## Hotfix (v3.0.3) Date 18.05.23
Include MIT license badges.

## Release (v3.0.2) Date 17.05.23
Fix issues with documentation.
* Fixed a formatting issue causing version history to run together.
* Fixed conflicting license use -- now all code is under MIT license.
* Fixed error in a reference entry in `paper/references.bib`.

## Hotfix (v3.0.1) Date 07.04.23
Hotfix for restoring Binder functionality.
* The repository now successfully builds on Binder again.
* Fixed CircleCI builds only retrieving files from master for testing; they now retrieve files from the commit being tested.

## Release (v3.0.0) Date 28.07.22

Create documentation oriented toward developers and reorganize libraries.
* Documented library functions.
* Condensed small variables and library functions that are only called by other library functions into fewer, larger ones.
* Fixed readthedocs badges.

## Release (v2.0.1) Date 30.06.22

* Improved notebook 06 by restoring the ability to "Run all cells below" with a button press instead of JupyterLab UI.
* Added another CI to test an environment generated more similarly to how Binder generates its environments.

## Release (v2.0.0) Date 24.06.22

Condensed many library functions.
* PR #25 combined many independent but similar functions and libraries using components.py.
* PR #25 also fixed an issue in which the "Reveal!" button in notebook #5 sent its output to the log instead of printing normally.
* Fixed some inconsistencies in date format between different documents -- all dates should now be dd.mm.yy

## Release (v1.0.4) Date 17.06.22

Author name update.

## Release (v1.0.3) Date 17.06.22

Addressed author comments regarding current paper. 
* Citation changes
* Table format
* Clarifications/corrections

## Release (v1.0.2) Date 09.06.22

More thorough documentation, including updates to the paper.
* PR #19 added coverage data and badges.
* PR #20 created documentation for the notebooks to be used for the workshop.
* PR #21 completed a draft of the paper for author comments period.

## Release (v1.0.1) Date 27.04.22 

* Merged branches to fix missing documentation updates.

## Release (v1.0.0) Date 27.04.22 

DM Workshop 1 released for RaCAS 2022.
* Many files were renamed to improve the organization of the repository.
* Added Code of Conduct and Issue Templates.
* PR #3 fixed with the provided environment that sometimes caused sliders not to update anything.
* PR #4 added testing to Travis-CI
* PR #17 added documents such as CONTRIBUTING, PULL_REQUEST_TEMPLATE, and these RELEASENOTES.

## Release (v0.0.1) Date 08.01.22

* Initial release! Repository created.


binder/python/w5.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


################################
########### Imports ############
################################
from sympy.utilities.lambdify import lambdify
import numpy as np
import matplotlib.pyplot as plt
from ipywidgets import interactive, fixed, FloatSlider, HBox, Layout, Button, Label, Output, VBox
from IPython.display import display
import code7814 as wi
import matplotlib.image as mpimg
import sys
sys.path.append('../../python/')          # Defining path
import dataPython as dp
#getting so that the halo component/total RC curve updates as bh widgets update


def h(r,Mbh,arraysize):
    Me=wi.scale*Mbh*arraysize #units: [bh/dot]*[Msun/bh] * [dot] * [Msun] = Msun --> changes as sliders are moved
    MMM=Me*wi.viso #units: unitless * [Msun] = Msun
    return np.sqrt(MMM*wi.GG/wi.rcut) #is it rc or just r? the model looks better with rcut 
    #units: ( [Msun]* [(km/s)^2*(kpc/Msun)] / kpc )^.5 = km/s                                                                

def t(r,Mbh,arraysize):
    return np.sqrt((wi.d)**2
               +(wi.b)**2
               +(h(r,Mbh,arraysize))**2
               +(wi.g)**2)
img = mpimg.imread('../python/thing.jpg') #import special snowflake ngc 6814, which has visual diameter about 27.6kpc\ 

# Define plotting function
def f(arraysize,Mbh):
    
    #fractions: now changing the number of dots plottd within each bracket, which changes as the slider 
    #variable ("arraysize") changes
    
    #and trim to the first x elements of the pre-caclated radius arrays for each bracket
    arraysize=int(arraysize) #units: dot
    re = wi.rrr[:arraysize]
    anglee=wi.angle[:arraysize]
    
    #plot for each bracket, two equations for each bracket
    y=wi.c_y+re*np.sin(anglee) #y coordinates
    x=wi.c_x+re*np.cos(anglee) #x coordinates
    
    
    
    #below: changing the physical size of the plotted dots as the bh mass slider changes. the if else statement is fancy
    #but not necessary
    
    Mbh=1.5*Mbh #judging by eye
    if .1<=Mbh<1:
        Mbh=Mbh**1.5
    elif 1<=Mbh<2:
        Mbh=Mbh**2
    else:
        Mbh=Mbh**3 #display size of each marker
        
    
    f, (ax1, ax2) = plt.subplots(1, 2, sharey=False)
    f.set_figheight(15)
    f.set_figwidth(35)
    
     # Define r, for some reason this isn't working in code7814.py, so put directly here:
    data = dp.getXYdata_wXYerr('../testing/7814/ngc7814data')
    r = np.asarray(data['xx'])
    z=r[0]
    rr=[0,z]
    ax1.scatter(x,y,color="r",marker='o',s=Mbh)
    ax1.imshow(img)
    ax1.axis('off')
    ax2.tick_params(axis='both', which='major', labelsize=30)
    ax2.plot(r,h(r,Mbh,arraysize),label=("Dark Matter Halo"),color='green')
    ax2.plot(rr,[0,h(z,Mbh,arraysize)[0]],color='green') #straight lining connecting halo curve to origin, for visual aesthetic
    ax2.errorbar(r,wi.v_dat,yerr=wi.v_err1,fmt='bo',label='Data')
    ax2.plot(r,wi.b,label=("Bulge"),color='orange')
    ax2.plot(rr,[0,wi.b[0]],color='orange') #straight line for bulge
    ax2.plot(r,wi.d,label=("Disk"),color='purple')
    ax2.plot(rr,[0,wi.d[0]],color='purple') #straight line for disk
    ax2.plot(r,wi.g,label=("Gas"),color='blue')
    ax2.plot(r,t(r,Mbh,arraysize),label=("Total Curve"),color='red')
    ax2.plot(rr,[0,t(z,Mbh,arraysize)[0]],color='red') #straight line for total curve
    ax2.set_xlim([wi.minkpc, 19])
    ax2.set_ylim([0, 400])
    ax2.set_title('NGC 7814',fontsize = 40)
    ax2.set_ylabel('Velocity [km/s]',fontsize = 30)
    ax2.set_xlabel('radius [kpc]',fontsize = 30)
    ax2.legend(loc='upper right',prop={'size': 20},ncol=3)
    ax2.text(.1, 310, 'How close can you get? \nA reduced $\chi^2$ less than 1 is considered a good fit',fontsize=30)
    ax = plt.gca()
    xticks = ax.xaxis.get_major_ticks() 
    xticks[0].label1.set_visible(False)
    residuals = wi.v_dat - t(wi.r,Mbh,arraysize)
    # Determining errors
    errors = wi.v_err1**2 #second term is inclination uncertainty
    # Chi squared
    chisquared = np.sum(residuals**2/errors**2)
    #chisquared = stats.chisquare(v_dat,totalcurve(r,M,bpref,dpref,rc,rho00,gpref))
    reducedchisquared = chisquared * (1/(len(wi.r)-6))
    
    props = dict(boxstyle='round', facecolor='white', alpha=0.5)
    ax2.text(0.4,360,r"Reduced $\chi^2$:"+'\n'+"{:.2f}".format(reducedchisquared),bbox=props,size=30)
    
   
    
style = {'description_width': 'initial'}
layout = {'width':'600px'}

################################
######## Define Sliders ########
################################

#mass of each black hole
Mbh = FloatSlider(min=wi.minmass, max=wi.maxmass, step=wi.minmass, 
                value=wi.start,
                description='Mass of each lil black holes [Msun]', 
                readout= True,
                readout_format='.1f', 
                orientation='horizontal', 
                style=style, layout=layout)

#number of projected black dots slider
arraysize = FloatSlider(min=wi.Min, max=wi.Max, step=wi.stepN, 
                value=wi.best_A, 
                description='Number of lil black holes multiplied by scale %.1e'%wi.scale, 
                readout= True,
                readout_format='.2d', 
                orientation='horizontal', 
                style=style, layout=layout)


def interactive_plot(f):
    interact = interactive(f, Mbh = Mbh, arraysize=arraysize,continuous_update=False)
    return interact

################################
########### Button #############
################################


# Button to revert back to Best Fit
button = Button(
    description="Best Fit",
    button_style='warning', # 'success', 'info', 'warning', 'danger' or ''
    icon='check')
out = Output()

def on_button_clicked(_):
    #display(Javascript('IPython.notebook.execute_cells_below()'))
    Mbh.value = wi.start
    arraysize.value=wi.best_A
button.on_click(on_button_clicked)

# displaying button and its output together
#VBox([button,out,interactive_plot(f)])


# In[ ]:




